### PR TITLE
Add the ability to specify main MANIFEST.MF attributes from configuration

### PR DESCRIPTION
**Pull request by glefloch**
Currently we can specify custom manifest section but we cannot set custom attributes in the main section (other than through a real `MANIFEST.MF` file). 
This adds a `quarkus.manifest.attributes` property that allows setting attributes. 

However, I don't know where to put tests... 

relates to #12889

### CODE REVIEW

**Comment by geoand** [comment]
👍🏼 

Can we also have a test for this?

**Comment by glefloch** [comment]
Sure @geoand, I just don't know where would be the best place for it?

**Comment by geoand** [comment]
Maybe in the maven and / or gradle tests?

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/ManifestConfig.java [hunk 1]
```diff
@@ -14,6 +14,15 @@
     @ConfigItem(defaultValue = "true")
     public boolean addImplementationEntries;
 
+    /**
+     * Custom manifest attributes to be added to the main section of the MANIFEST.MF file.
+     * An example of the user defined property:
+     * quarkus.package.manifest.attributes.{Entry-key1}={Value1}
+     * quarkus.package.manifest.attributes.{Entry-key2}={Value2}
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/ManifestConfig.java:21]:
```suggestion
     * <pre>
     * quarkus.package.manifest.attributes."Entry-key1"=value1
     * quarkus.package.manifest.attributes."Entry-key2"=value2
     * </pre>
```
Having the curly braces can be confusing as we might think they are required for things to work.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java [hunk 2]
```diff
@@ -1333,15 +1333,22 @@ private void generateManifest(FileSystem runnerZipFs, final String classPath, Pa
...
+
+        if (config.manifest.attributes.size() > 0) {
+            for (Map.Entry<String, String> attribute : config.manifest.attributes.entrySet()) {
+                attributes.putIfAbsent(attribute.getKey(), attribute.getValue());
+            }
+        }
         if (attributes.containsKey(Attributes.Name.CLASS_PATH)) {
             log.warn(
-                    "Your MANIFEST.MF already defined a CLASS_PATH entry. Quarkus has overwritten this existing entry.");
+                    "A CLASS_PATH entry was already defined in your MANIFEST.MF or using the property quarkus.manifest.attributes.\"Class-Path\". Quarkus has overwritten this existing entry.");
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java:1344]:
```suggestion
                    "A CLASS_PATH entry was already defined in your MANIFEST.MF or using the property quarkus.package.manifest.attributes.\"Class-Path\". Quarkus has overwritten this existing entry.");
```

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java [hunk 3]
```diff
@@ -1333,15 +1333,22 @@ private void generateManifest(FileSystem runnerZipFs, final String classPath, Pa
...
         }
         attributes.put(Attributes.Name.CLASS_PATH, classPath);
         if (attributes.containsKey(Attributes.Name.MAIN_CLASS)) {
             String existingMainClass = attributes.getValue(Attributes.Name.MAIN_CLASS);
             if (!mainClassName.equals(existingMainClass)) {
-                log.warn("Your MANIFEST.MF already defined a MAIN_CLASS entry. Quarkus has overwritten your existing entry.");
+                log.warn(
+                        "A MAIN_CLASS entry was already defined in your MANIFEST.MF or using the property quarkus.manifest.attributes.\"Main-Class\". Quarkus has overwritten your existing entry.");
```
Comment (gsmet) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java:1351]:
```suggestion
                        "A MAIN_CLASS entry was already defined in your MANIFEST.MF or using the property quarkus.package.manifest.attributes.\"Main-Class\". Quarkus has overwritten your existing entry.");
```

**Review by gsmet — CHANGES_REQUESTED** [review]
Spotted a few minor issues.

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/ManifestConfig.java (hunk 1 above)
Comment (glefloch) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/ManifestConfig.java]: I removed curly braces from the manifest sections property comment too.

**Comment by glefloch** [comment]
Thanks for reviews, I updated the code and added a test in the gradle integration test module.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]
We already have a module specifically for this type of test:

https://github.com/quarkusio/quarkus/blob/01bee6969efe374d4489d35189c65f34106cea87/integration-tests/packaging/src/test/java/io/quarkus/maven/CustomManifestSectionsThinJarTest.java#L20

I know the suggestion was to add a gradle based test but IMHO it makes much more sense to just add to the existing integration tests, as this does not really have anything to do with gradle.

**Review by stuartwdouglas — APPROVED** [review]
I just pushed a commit that changes the test to be consistent.

**Comment by geoand** [comment]
Thanks Stuart

**Review by geoand — APPROVED** [review]

**Comment by glefloch** [comment]
Thanks a lot @stuartwdouglas

**Comment by glefloch** [comment]
CI failure does not seem related.

**Review by gsmet — APPROVED** [review]

**Pull request merged**